File: src/day08/Day08.kt
package day08

import java.io.File

const val FILENAME = "src/day08/Input.txt"

val instructions = File(FILENAME).readLines().toTypedArray()

fun part01(runningPart02: Boolean = false): Int {
    var accumulator = 0
    var reachedEnd = false
    val executed = BooleanArray(instructions.size)
    var i = 0
    while (i < instructions.size) {
        if (i == instructions.size - 1)
            reachedEnd = true
        if (executed[i])
            break
        when {
            instructions[i].contains("nop") -> {
                executed[i] = true
                i++
            }
            instructions[i].contains("acc") -> {
                accumulator += instructions[i].substring(4).toInt()
                executed[i] = true
                i++
            }
            instructions[i].contains("jmp") -> {
                val jumpTo = i + instructions[i].substring(4).toInt()
                executed[i] = true
                i = jumpTo
            }
        }
    }
    // If running only part01, return accumulator as it is
    if (!runningPart02)
        return accumulator
    else {
        // If running again for part02, return accumulator only if the final instruction
        // is reached. Else return -1, as final instruction has not been reached.
        if (reachedEnd)
            return accumulator
        else
            return -1
    }
}

fun part02(): Int {
    for (i in instructions.indices) {
        if (instructions[i][0] != 'a') {
            val beforeReplace = instructions[i]
            if (instructions[i][0] == 'n')
                instructions[i] = instructions[i].replace("nop", "jmp")
            else if (instructions[i][0] == 'j')
                instructions[i] = instructions[i].replace("jmp", "nop")

            val accumulator = part01(runningPart02 = true)
            if (accumulator > 0)
                return accumulator
            else
                instructions[i] = beforeReplace
        }
    }
    return -1
}

fun main() {
    println("Advent of Code 2020, Day 08")
    println(part01())
    println(part02())
}

File: day08/day08.go
package day08

import (
	"advent-of-code-2020/utils"
	"strconv"
	"strings"
)

const filename = "./inputs/day08.txt"

var instructions = utils.ReadLines(filename)

func getAccumulator(part02 bool) int {
	accumulator := 0
	reachedEnd := false
	executed := make([]bool, len(instructions))
	i := 0
	for i < len(instructions) {
		if i == len(instructions)-1 {
			reachedEnd = true
		}
		if executed[i] {
			break
		}
		switch {
		case strings.Contains(instructions[i], "nop"):
			executed[i] = true
			i++
		case strings.Contains(instructions[i], "acc"):
			num, _ := strconv.Atoi(instructions[i][4:])
			accumulator += num
			executed[i] = true
			i++
		case strings.Contains(instructions[i], "jmp"):
			num, _ := strconv.Atoi(instructions[i][4:])
			jumpTo := i + num
			executed[i] = true
			i = jumpTo
		}
	}

	// If running only part01, return accumulator as it is
	if !part02 {
		return accumulator
	}

	// If running for part02, return accumulator only if the final instruction
	// is reached, else, return -1, as final instruction has not been reached.
	if reachedEnd {
		return accumulator
	} else {
		return -1
	}
}

func part01() int {
	return getAccumulator(false)
}

func part02() int {
	for i := range instructions {
		instruction := []rune(instructions[i])
		if instruction[0] != 'a' {
			beforeReplace := instructions[i]
			if instruction[0] == 'n' {
				instructions[i] = strings.ReplaceAll(instructions[i], "nop", "jmp")
			} else if instruction[0] == 'j' {
				instructions[i] = strings.ReplaceAll(instructions[i], "jmp", "nop")
			}

			accumulator := getAccumulator(true)
			if accumulator > 0 {
				return accumulator
			} else {
				instructions[i] = beforeReplace
			}
		}
	}
	return -1
}

func Main() (int, func() int, func() int) {
	return 8, part01, part02
}
